config/configmapprovider/provider.go

```diff
@@ -48,7 +58,14 @@ type Provider interface {
 	//
 	// If ctx is cancelled should return immediately with an error.
 	// Should never be called concurrently with itself or with Shutdown.
-	Retrieve(ctx context.Context, watcher WatcherFunc) (Retrieved, error)
+	Retrieve(ctx context.Context, url *url.URL, watcher WatcherFunc) (Retrieved, error)
+
+	// Scheme returns the implemented scheme name (see https://datatracker.ietf.org/doc/html/rfc3986#section-3.1)
+	// by this Provider.
+	//
+	// Implementation must make sure to not conflict with well known scheme names,
+	// see https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml
+	Scheme() string
```

Comment:
Is there any provision for a provider supporting multiple schemes, or multiple providers supporting the same scheme?  For instance, the prototype S3 provider I created could use either `s3://` or `http://` URLs.

Reply:
Hmm.. There is a problem with that "http" cannot be restricted to only "s3", and if you do register the "http" for s3 an alternative that loads from any HTTP endpoint cannot exists.

Reply:
Yup, not sure there's a great solution to that other than not supporting HTTP URIs for S3 configs.  Maybe something could be done with patterns instead of schemes, or a chain of responsibility ending in a generic HTTP loader that could give other providers a chance to determine if a given URI should be handled.

Reply:
But a generic HTTP/GET provider should work for the s3 case as well correct? Does not need to be something special.

Reply:
For objects that are public, it would.  The prototype I built instead decomposed the URI to extract the information needed to use the S3 client and thus could be used for any object, even with the HTTP syntax.  This would be necessary if the object were in a different region as the [`s3://` syntax](https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html#accessing-a-bucket-using-S3-format) does not support specifying a region.

Reply:
Not sure what you are arguing for though:
1. I think one provider is responsible for only one "scheme" ("source_id");
2. If a well known URL scheme is defined for that name, then the provider MUST follow the RFCs that defines it.